src/unicode/capi.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * This file is part of the "libunicode" project
+ *   Copyright (c) 2021 Christian Parpart <christian@parpart.family>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if !defined(__cplusplus)
+extern "C" {
+#endif
+
+/// UTF-8 character or 8bit segment of an UTF-8 character.
+typedef char u8_char_t;
+
+/// UTF-32 codepoint.
+typedef uint32_t u32_char_t;
+
+/**
+ * Counts the number of grapheme clusters for given sequence of codepoints.
+ *
+ * Use this function to determine the number of
+ * user perceived characters (grapheme clusters).
+ *
+ * @param codepoints   pointer to the first codepoint.
+ * @param n            number of codepoints to count the grapheme clusters for.
+ *
+ * @return number of user perceived characters (grapheme clusters) counted
+ *         in [codepoints, codepoints+n).
+ */
+int u32_gc_count(u32_char_t const* codepoints, size_t n);
+int u8_gc_count(u8_char_t const* codepoints, size_t n);
+
+/**
+ * Computes the display width for given sequence of codepoints,
+ * respecting grapheme cluters, and modifiers.
+ *
+ * @param codepoints  pointer to first codepoint
+ * @param n           number of codepoints
+ * @param allowMod    booelan to indicate whether width modifications
+ *                    (e.g. via VS15/VS16) are allowed or not.
+ *
+ * Use this function to determine how many terminal grid cells a
+ * string of codepoints should occupy when being rendered.
+ */
+int u32_gc_width(u32_char_t const* codepoints, size_t n, int allowMod);
```

Comment:
Again very nice to break it down to such a simple API. For a release imho `allowMod` should be documented to its full extend (list all modifications, that can be activated/used during this).

Reply:
`bool allowMod` does exist because to preserve current application behavior (i.e. you don't wanna break them) you should set it to false. If you want to conform to the spec, you should set it to true. A terminal emulator could let the user decide how to act, or even create a VT sequence (`SM/RM ?modeNum`) to enable/disable conformance (my recommendation).

I can make that a binary enum though, I absolutely agree here. And also extending documentation should be done. thanks. :)

Reply:
Ah wait - so the mod num here refers to some SM/RM setting of a terminal? Maybe I misunderstood - isnt that too specific to TEs in the end?

Reply:
No you did understand correctly! I was giving a case for why you want that configurable.

Reply:
Ah ok. Can you give me examples, what you want to see here as modifications? Just wondering, if this will impose functionality frictions later on between apps & TEs (in the sense of - if a terminal app uses mod XY here, the TE has to support it as part of its infrastructure design).

Reply:
Hmm... without trying to specc out some core Unicode features for the TE land, my idea of my local draft spec has the following:

* `DA1` advertises that the TE does support unicode core and therefore the following
* `SM?` / `RM?` get an additional mode number to explicitly  unicode core conformance (i.e. grapheme cluster processing AND the fact that the variation selector can change the display width of a character).

For that to work conveniently on the TE/TUI side, it should be possible to calculate the display width  in the old way as well as in the conforming way.